### ISSUE TITLE
interaction with mesh surface

### ISSUE THREAD

**Issue by Ali-RAFIEE** [siconos/siconos#374]
Hi everyone,

I have some questions about interaction between a mesh surface and polyhedron shapes.
in fact I use this below simple python code to do that, and in this model I have just a simple cube block and a mesh surface in stl format.

questions:
1-my stl mesh file (attached zip file) is in triangulated mesh form, with normal of each triangle surface towards to impacting block. is this way correct?
2- is my code structure generally correct?
3- should I use (or is it better) to use another configuration in my python code to model this type of simulation? eg using Bullet Option?
4- when I change my block shape to a more complex shape (even in convex shape), I see a large interpenetration between the block and the slope surface.

I think something in my code is not well defined.

Thanks a lot

[slope2.zip](https://github.com/siconos/siconos/files/5728901/slope2.zip)
=============================================================
```
from siconos.mechanics.collision.tools import Contactor
from siconos.io.mechanics_run import MechanicsHdf5Runner
from siconos.mechanics.collision.convexhull import ConvexHull

import siconos.numerics as sn
import siconos.kernel as sk
density=2000

# Creation of the hdf5 file for input/output

with MechanicsHdf5Runner() as io:
    vertices=[(-2.5,7.829999923706,17.375),(-2.5,7.829999923706,18.625),(-2.5,8.829999923706,17.375),(-2.5,8.829999923706,18.625),(-1.5,7.829999923706,17.375),(-1.5,7.829999923706,18.625),(-1.5,8.829999923706,17.375),(-1.5,8.829999923706,18.625)]
    io.add_convex_shape("bloc-1", vertices)
    ch = ConvexHull(vertices)
    inertia,volume=ch.inertia(ch.centroid())
    io.add_object("Block-1", [Contactor("bloc-1")], translation=[0, 0, 0] , mass=volume*density,inertia=inertia*density)


    io.add_mesh_from_file('Slope', "slope2.stl", scale=1.0,insideMargin=0.0, outsideMargin=0.0)
    
    io.add_object('ground', [Contactor('Slope')],translation=[0, 0, 0.0])
    
    io.add_Newton_impact_friction_nsl('contact', e=0.25, mu=0.7)
    
    
# Create solver options
options = sk.solver_options_create(sn.SICONOS_FRICTION_3D_NSGS)
options.iparam[sn.SICONOS_IPARAM_MAX_ITER] = 10000
options.iparam[sn.SICONOS_FRICTION_3D_NSGS_FREEZING_CONTACT] = 100
options.dparam[sn.SICONOS_DPARAM_TOL] = 1e-3


T=20.0
hstep=0.0001

with MechanicsHdf5Runner(mode='r+') as io:
    io.run(with_timer=False,
           explode_Newton_solve=False,
           t0=0,
           T=T,
           h=hstep,
           multipoints_iterations=True,
           theta=1.0,
           Newton_max_iter=50,
           solver_options=options,
           output_frequency=50,
           )
```
=============================================================

**Comment by vacary** [comment]
On my installation, your file is not producing simulation results due to vtk error in your stl file.

Could you provide me with an image of the scene ?

**Comment by ali145** [comment]
Hi Vincent,
Happy new year 2021.

I am running this code without any problem, and I can get the output file in hdf5 format, and it is convertible to vtk format, by attachments, you can find the hdf5 and also the image from paraview.

Thanks a lot
Ali

![vtk-image-test](https://user-images.githubusercontent.com/25464988/103519124-43c42400-4e75-11eb-8a05-7993a6b5f847.png)

[traject-cube.zip](https://github.com/siconos/siconos/files/5763894/traject-cube.zip)

**Comment by vacary** [comment]
Ok, I have a problem with vtk on My MacOs Catalina. I will fix it and I come back to you.

**Comment by vacary** [comment]
Hi,
On my system (MacOs Catalina), I have to fix a lot of issue due to h5py 3.1.0 and I am forces to use it by the new Xcode :-(
Finally, I fix this problem and I was able to have a look to your problem.
The issue comes from the default values of the margin (inside and outside) in Bullet3. I fixed the problem on my computer:
by replacing the line
```
io.add_convex_shape("bloc-1", vertices)
```
by
```
io.add_convex_shape("bloc-1", vertices, insideMargin=0.0, outsideMargin=0.0)
```
Let me know if it works on your side ?

**Comment by Ali-RAFIEE** [comment]
Hi Vincent,
Thanks for your time, but I still saw the problem of contact detection between this mesh surface and the simple block even after doing your suggested modification.

as you can see, in this attached image clearly, the block passe partially through the surface.
![Problem-contact-detection](https://user-images.githubusercontent.com/25464988/104424289-78279680-557f-11eb-8b72-7f63536753d7.png)

**Comment by vacary** [comment]
What is the version of bullet ? Do you use is in double precision ?

**Comment by Ali-RAFIEE** [comment]
libbullet-dev (2.87+dfsg-2)

**Comment by Ali-RAFIEE** [comment]
for installing bullet in double precision I'm trying again to reinstall it in this way:

git https://github.com/bulletphysics/bullet3

cd bullet3/build3

sudo ./premake4_linux64 --double gmake

cd gmake
sudo make

I hope it will work this time.

**Comment by Ali-RAFIEE** [comment]
finally I could not install libbullet in double precision version.

I do not know how can I do this for Ubuntu 18.04.

**Comment by Ali-RAFIEE** [comment]
I have this error message for installation of Siconos after installing bullet double precision with this command:

./build_cmake_pybullet_double.sh
cd build_cmake
make

the bullet is installed without any problem but for Siconos installation I received the below error message.

I do not know why?

=====
[ 95%] Linking CXX executable src/collision/bullet/test/Contact2dTest
libsiconos_mechanics.so.6.0.0: undefined reference to `btGImpactMeshShape::calculateLocalInertia(float, btVector3&) const'
libsiconos_mechanics.so.6.0.0: undefined reference to `btHeightfieldTerrainShape::btHeightfieldTerrainShape(int, int, void const*, float, float, float, int, PHY_ScalarType, bool)'
libsiconos_mechanics.so.6.0.0: undefined reference to `btConeShape::btConeShape(float, float)'
libsiconos_mechanics.so.6.0.0: undefined reference to `btConvexHullShape::btConvexHullShape(float const*, int, int)'
libsiconos_mechanics.so.6.0.0: undefined reference to `btHeightfieldTerrainShape::calculateLocalInertia(float, btVector3&) const'
libsiconos_mechanics.so.6.0.0: undefined reference to `btCollisionShape::getBoundingSphere(btVector3&, float&) const'
libsiconos_mechanics.so.6.0.0: undefined reference to `btCapsuleShape::btCapsuleShape(float, float)'
libsiconos_mechanics.so.6.0.0: undefined reference to `btTriangleIndexVertexArray::btTriangleIndexVertexArray(int, int*, int, int, float*, int)'
libsiconos_mechanics.so.6.0.0: undefined reference to `btCollisionShape::getContactBreakingThreshold(float) const'
libsiconos_mechanics.so.6.0.0: undefined reference to `btConvexHullComputer::compute(void const*, bool, int, int, float, float)'
collect2: error: ld returned 1 exit status
mechanics/CMakeFiles/Contact2dTest.dir/build.make:111: recipe for target 'mechanics/src/collision/bullet/test/Contact2dTest' failed
make[2]: *** [mechanics/src/collision/bullet/test/Contact2dTest] Error 1
CMakeFiles/Makefile2:6419: recipe for target 'mechanics/CMakeFiles/Contact2dTest.dir/all' failed
make[1]: *** [mechanics/CMakeFiles/Contact2dTest.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
===================================

**Comment by vacary** [comment]
No idea, ask the bullet guys with an issue

**Comment by Ali-RAFIEE** [comment]
for siconos I execute these commands in build folder:

cmake .. -DWITH_TESTING=ON -DWITH_PYTHON_WRAPPER=ON -DWITH_DOCUMENTATION=ON -DWITH_BULLET=ON -DWITH_XML=ON -DWITH_RENDERER=ON

make

sudo make install

**Comment by bremond** [comment]
By default, siconos uses the bullet simple precision API.

For the double precision API, there is a flag that must be added at
cmake time: -DBULLET_USE_DOUBLE_PRECISION=1

(This worked a long time ago and it is untested).

Note that a make clean is mandatory before re-compilation.

**Comment by Ali-RAFIEE** [comment]
Thanks a lot
it seems to work.

**Comment by Ali-RAFIEE** [comment]
after reinstalling siconos with double precision for Bullet, when I run the above example, I have this error message.
I think it's related to mesh surface that I have in my model.


=======================
python3 traject-cube.py
No module named siconos.fclib.
[io.mechanics] setup model simulation ...
[io.mechanics] 
[io.mechanics] Simulation will run from time 0.0000 to 20.0000s, step 1 to step 200001 (h=0.0001, times=[?,?])
[io.mechanics] 
[io.mechanics] import scene ...
[io.mechanics] import dynamical systems ...
[io.mechanics] Import  dynamic or static object number  1 from initial state
[io.mechanics]                 object name    Block-1
[io.mechanics] Import  dynamic or static object number  1 from initial state
[io.mechanics]                 object name    Block-1
[io.mechanics] Import  dynamic or static object number  -1 from initial state
[io.mechanics]                 object name    ground
[io.mechanics] Import  dynamic or static object number  -1 from initial state
[io.mechanics]                 object name    ground
Traceback (most recent call last):
  File "traject-cube.py", line 57, in <module>
    output_frequency=50,
  File "/usr/local/lib/python3.6/dist-packages/siconos/io/mechanics_run.py", line 2333, in run
    self.import_scene(t0, body_class, shape_class, face_class, edge_class)
  File "/usr/local/lib/python3.6/dist-packages/siconos/io/mechanics_run.py", line 1656, in import_scene
    birth=False)
  File "/usr/local/lib/python3.6/dist-packages/siconos/io/mechanics_run.py", line 1539, in import_object
    number=self.instances()[name].attrs['id'])
  File "/usr/local/lib/python3.6/dist-packages/siconos/io/mechanics_run.py", line 1032, in import_bullet_object
    shp = self._shape.get(c.shape_name)
  File "/usr/local/lib/python3.6/dist-packages/siconos/io/mechanics_run.py", line 471, in get
    tmpf[0].write(data.decode("utf-8"))
AttributeError: 'str' object has no attribute 'decode'

**Comment by vacary** [comment]
I have to fix this problem due to an update for h5py 3.1.0
